core/services/ocrcommon/transmitter.go

```diff
@@ -203,3 +209,108 @@ func (t *ocr2FeedsTransmitter) forwarderAddress(ctx context.Context, eoa, ocr2Ag
 
 	return forwarderAddress, nil
 }
+
+type saveIdempotencyKey func(uuid uuid.UUID, uid *big.Int) error
+
+type ocr3AutomationTransmitter struct {
+	transmitter
+	ChainType config.ChainType
+	lggr logger.Logger
+	saveIdempotencyKey
+}
+
+// NewOCR3AutomationTransmitter creates a new eth transmitter
+func NewOCR3AutomationTransmitter(
+	txm txManager,
+	fromAddresses []common.Address,
+	gasLimit uint64,
+	effectiveTransmitterAddress common.Address,
+	strategy types.TxStrategy,
+	checker txmgr.TransmitCheckerSpec,
+	chainID *big.Int,
+	keystore roundRobinKeystore,
+	chainType config.ChainType,
+	saveIdempotencyKey func(uuid uuid.UUID, uid *big.Int) error,
+	lggr logger.Logger,
+) (Transmitter, error) {
+	// Ensure that a keystore is provided.
+	if keystore == nil {
+		return nil, errors.New("nil keystore provided to transmitter")
+	}
+
+	return &ocr3AutomationTransmitter{
+		transmitter: transmitter{
+			txm:                         txm,
+			fromAddresses:               fromAddresses,
+			gasLimit:                    gasLimit,
+			effectiveTransmitterAddress: effectiveTransmitterAddress,
+			strategy:                    strategy,
+			checker:                     checker,
+			chainID:                     chainID,
+			keystore:                    keystore,
+		},
+		ChainType: chainType,
+		saveIdempotencyKey: saveIdempotencyKey,
+		lggr: lggr,
+	}, nil
+}
+
+func (t *ocr3AutomationTransmitter) CreateEthTransaction(ctx context.Context, toAddress common.Address, payload []byte, txMeta *txmgr.TxMeta) error {
+	roundRobinFromAddress, err := t.keystore.GetRoundRobinAddress(ctx, t.chainID, t.fromAddresses...)
+	if err != nil {
+		return errors.Wrap(err, "skipped OCR transmission, error getting round-robin address")
+	}
+
+	var id uuid.UUID
+	var key string
+	var keyPtr *string
+	if t.ChainType == config.ChainScroll || t.ChainType == config.ChainZkEvm || t.ChainType == config.ChainZkSync {
+		if txMeta != nil && txMeta.UpkeepID != nil {
+			id, err = uuid.NewRandomFromReader(bytes.NewReader([]byte(*txMeta.UpkeepID + time.Now().String())))
+			if err != nil {
+				t.lggr.Errorf("failed to create UUID from %s", *(txMeta.UpkeepID))
+			} else {
+				uid, ok := new(big.Int).SetString(*(txMeta.UpkeepID), 10)
+				if !ok {
+					t.lggr.Errorf("failed to convert upkeep ID %s to big int", *(txMeta.UpkeepID))
+				} else {
+					err = t.saveIdempotencyKey(id, uid)
+					if err != nil {
+						t.lggr.Errorf("failed to save idempotency key %s due to %s", id.String(), err.Error())
+					} else {
+						key = id.String()
+						keyPtr = &key
+					}
+				}
+			}
+		} else {
+			t.lggr.Errorf("failed to retrieve upkeep ID from tx meta")
+		}
+	}
+
+	_, err = t.txm.CreateTransaction(ctx, txmgr.TxRequest{
+		IdempotencyKey:   keyPtr,
```

Comment:
tx with the same idempotency key will not be sent again
1 key means 1 onchain tx